File: clients/ios/StellarChat/StellarChat/Models/KeyAttestation.swift
import CryptoKit
import Foundation

/// Binds a BLS12-381 group membership key to a secp256k1 Nostr identity key.
///
/// Per SEP-XXXX §1.1, the attestation proves that the holder of the Nostr signing key
/// also controls the BLS key used in the group Merkle tree. The signature covers
/// `SHA-256("SEP-XXXX:key-binding" || bls_pubkey)`.
///
/// The spec defines Ed25519 signatures for Stellar address binding; this implementation
/// uses secp256k1 Schnorr signatures for Nostr identity binding.
struct KeyAttestation: Codable, Equatable {
    let blsPubkey: Data      // 48 bytes, compressed G1 point
    let nostrPubkey: Data    // 32 bytes, secp256k1 x-only public key
    let signature: Data      // 64 bytes, Schnorr signature

    static func bindingMessage(blsPubkey: Data) -> Data {
        var hasher = SHA256()
        hasher.update(data: Data("SEP-XXXX:key-binding".utf8))
        hasher.update(data: blsPubkey)
        return Data(hasher.finalize())
    }
}

File: clients/android/StellarChat/app/src/main/java/chat/onym/android/crypto/KeyAttestation.kt
package chat.onym.android.crypto

import java.security.MessageDigest

/**
 * BLS ↔ Stellar Ed25519 key binding (SEP-XXXX §1.1).
 * Proves that the same entity controls both the BLS group membership key
 * and the Stellar on-chain identity.
 */
data class KeyAttestation(
    val blsPubkey: ByteArray,     // 48 bytes, compressed G1 point
    val ed25519Pubkey: ByteArray, // 32 bytes, Stellar Ed25519 public key
    val signature: ByteArray      // 64 bytes, Ed25519 signature
) {
    companion object {
        /** Compute the binding message: SHA256("SEP-XXXX:key-binding" || blsPubkey). */
        fun bindingMessage(blsPubkey: ByteArray): ByteArray {
            val digest = MessageDigest.getInstance("SHA-256")
            digest.update("SEP-XXXX:key-binding".toByteArray())
            digest.update(blsPubkey)
            return digest.digest()
        }

        /** Create an attestation binding a BLS public key to a Stellar Ed25519 key. */
        fun create(keyManager: KeyManager): KeyAttestation {
            val blsPub = keyManager.blsPublicKey()
            val message = bindingMessage(blsPub)
            val signature = keyManager.stellarSign(message)
            return KeyAttestation(
                blsPubkey = blsPub,
                ed25519Pubkey = keyManager.stellarPublicKey,
                signature = signature
            )
        }

        /** Verify that the attestation signature is valid. */
        fun verify(attestation: KeyAttestation): Boolean {
            val message = bindingMessage(attestation.blsPubkey)
            return KeyManager.verifyEd25519(
                publicKey = attestation.ed25519Pubkey,
                message = message,
                signature = attestation.signature
            )
        }
    }

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other !is KeyAttestation) return false
        return blsPubkey.contentEquals(other.blsPubkey) &&
                ed25519Pubkey.contentEquals(other.ed25519Pubkey) &&
                signature.contentEquals(other.signature)
    }

    override fun hashCode(): Int {
        var result = blsPubkey.contentHashCode()
        result = 31 * result + ed25519Pubkey.contentHashCode()
        result = 31 * result + signature.contentHashCode()
        return result
    }
}
